proposals/2023/04-21_build-init-log.md

```diff
@@ -0,0 +1,292 @@
+
+So, relying on a magic `"init"` string is surprising and problematic. Relying on `"#init"` as a magic string on step.Image is only marginally better.
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+### Types
+
+No changes required for `database.Log` or `library.Log` structs since they are already associated with a `BuildID`.
+Validation, however, will need to allow for both `StepID` and `ServiceID` to be `NULL`.
+
+Drop this check in `Validate()` in `database/log.go` since no `StepID` and no `ServiceID` means we have a `Build` `Log`:
+<!-- https://github.com/go-vela/types/blob/main/database/log.go#L129-L132 -->
+```
+	// verify the has StepID or ServiceID field populated
+	if l.StepID.Int64 <= 0 && l.ServiceID.Int64 <= 0 {
+		return ErrEmptyLogStepOrServiceID
+	}
+```
+<!-- https://github.com/go-vela/types/blob/main/database/log_test.go#L298-L299
+And this needs to be failure: false
+```
+		{ // no service_id or step_id set for log
+			failure: true,
+```
+-->
+
+No changes are needed in the `pipeline` or the `yaml` layers because there is not a `pipeline.Log` nor a `yaml.Log`, so the relationship between `Build` and `Log` does not need to be represented.
+
+### Server
+
+#### Server - Database
+
+Adjust `CreateLog`, `UpdateLog`, and `DeleteLog` to handle the build `Log` (adding a `logger.Tracef` entry and an `Error` specific to the Build Log).
+
+Add `database/log/get_build.go` with:
+```
+// GetLogForBuild gets a log by build ID from the database.
+func (e *engine) GetLogForBuild(b *library.Build) (*library.Log, error) {
+```
+
+See if we can make the build log sort before the step logs in `ListLogsForBuild`:
+<!-- https://github.com/go-vela/server/blob/main/database/log/list_build.go#L37-L40 -->
+```
+	err = e.client.
+		Table(constants.TableLog).
+		Where("build_id = ?", b.GetID()).
+		Order("step_id ASC").
+```
+
+Sqlite sorts with `NULLS FIRST` by default, and Postgres sorts with `NULLS LAST` (see: [How Are NULLs Sorted by Default?](https://learnsql.com/blog/how-to-order-rows-with-nulls/)).
+So this query is inconsistent between databases. And, the service logs returned by this query are not sorted. It would be nice to have this method return the logs sorted in a consistent manner.
+I would like to see the Build Log, then Step Logs sorted by step_id, then Service Logs sorted by service_id. Any suggestions on how to do that with gorm?
```

Comment:
Do you know if GORM supports multiple `ORDER BY` statements? i.e.

```go
	err = e.client.
		Table(constants.TableLog).
		Where("build_id = ?", b.GetID()).
		Order("step_id ASC").
		Order("service_id ASC").
```

I could see a scenario where instead we have to use code like:

```go
	err = e.client.
		Table(constants.TableLog).
		Where("build_id = ?", b.GetID()).
		Order("step_id ASC, service_id ASC").
```

Based off my brief reviews, I'm seeing docs that suggest both Postgres and Sqlite support this:

* https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-order-by/
* https://www.sqlitetutorial.net/sqlite-order-by/

Reply:
https://gorm.io/docs/query.html#Order

Yep it supports multiple order statements.

Reply:
How about this? I believe this will sort the logs as [init, steps..., services...] and the sort order will be consistent for both sqlite and postgres.

```go
	err = e.client.
		Table(constants.TableLog).
		Where("build_id = ?", b.GetID()).
		Order("(step_id IS NULL AND service_id IS NULL) DESC"). // the build init log
		Order("step_id ASC NULLS LAST").
		Order("service_id ASC").
```

If that proves problematic, then we could add a computed column:

```go
type Log struct {
	ID        sql.NullInt64 `sql:"id"`
	BuildID   sql.NullInt64 `sql:"build_id"`
	RepoID    sql.NullInt64 `sql:"repo_id"`
	ServiceID sql.NullInt64 `sql:"service_id"`
	StepID    sql.NullInt64 `sql:"step_id"`
	Data      []byte        `sql:"data"`
	IsInit    sql.Bool      `sql:"is_init"    gorm:"->"`
}
```
```sql
CREATE TABLE
IF NOT EXISTS
logs (
	id            SERIAL PRIMARY KEY,
	build_id      INTEGER,
	repo_id       INTEGER,
	service_id    INTEGER,
	step_id       INTEGER,
	data          BYTEA,
	is_init       BOOLEAN GENERATED ALWAYS AS (step_id IS NULL AND service_id IS NULL) STORED,
	UNIQUE(step_id),
	UNIQUE(service_id)
);
```
which works in both [postgres](https://www.postgresql.org/docs/current/ddl-generated-columns.html) and [sqlite](https://www.sqlite.org/gencol.html). Postgres only supports `STORED`, but in sqlite a `STORED` column can't be added to an existing table, so the sqlite migration would have to add the column as `VIRTUAL` instead.

That would lead to this query:

```go
	err = e.client.
		Table(constants.TableLog).
		Where("build_id = ?", b.GetID()).
		Order("is_init DESC").
		Order("step_id ASC NULLS LAST").
		Order("service_id ASC").
```

Reply:
After further thought, I think we do need an `is_init` boolean column. Instead of delegating that calculation to the database, we'll just calculate it when creating the objects, that way it will be accessible in `log.Validate`.

So:

```sql
CREATE TABLE
IF NOT EXISTS
logs (
	id            SERIAL PRIMARY KEY,
	build_id      INTEGER,
	repo_id       INTEGER,
	service_id    INTEGER,
	step_id       INTEGER,
	is_init       BOOLEAN,
	data          BYTEA,
	UNIQUE(step_id),
	UNIQUE(service_id),
	CHECK(is_init IS NOT FALSE),
	UNIQUE(build_id, is_init)
);
```

Then `log.Validate` will ensure that one of `is_init`, `step_id`, and `service_id` are not `NULL`.
And we'll only ever set `is_init` to `TRUE` if both `step_id` and `service_id` are `NULL`. And, we never set `is_init` to `FALSE` to allow the composite unique constraint to ensure that each build has only one init log.

Reply:
Oh. I just asked someone on the DB team at work, and he said this should work:
```sql
CREATE TABLE
IF NOT EXISTS
logs (
	id            SERIAL PRIMARY KEY,
	build_id      INTEGER,
	repo_id       INTEGER,
	service_id    INTEGER,
	step_id       INTEGER,
	data          BYTEA,
	UNIQUE(build_id, step_id, service_id)
);
```